modules/gateway/rpc.go

```diff
@@ -93,17 +98,37 @@ func (g *Gateway) RegisterConnectCall(name string, fn modules.RPCFunc) {
 // corresponding RPCFunc from g.initRPCs. Future connections to peers will not
 // trigger the RPC to be called on them.
 func (g *Gateway) UnregisterConnectCall(name string) {
-	id := g.mu.Lock()
-	defer g.mu.Unlock(id)
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	if _, ok := g.initRPCs[name]; !ok {
 		build.Critical("ConnectCall not registered: " + name)
 	}
 	delete(g.initRPCs, name)
 }
 
-// listenPeer listens for new streams on a peer connection and serves them via
+// threadedListenPeer listens for new streams on a peer connection and serves them via
 // threadedHandleConn.
-func (g *Gateway) listenPeer(p *peer) {
+func (g *Gateway) threadedListenPeer(p *peer) {
+	if g.threads.Add() != nil {
+		return
+	}
+	defer g.threads.Done()
+
+	// Disconnect from the peer when the stop signal is received.
+	if g.threads.Add() != nil {
+		return
+	}
+	go func() {
```

Comment:
Does this code take into account situations where you disconnect from the peer before `Close()` is called? What happens to this goroutine if your peer leaves early?

I think right now it's a memory leak.